File: VideoGameClient/src/app/core/services/video-game.service.ts
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { GameType } from '../models/enums';
import { Platform } from '../models/platform';
import { Publisher } from '../models/publisher';
import { VideoGame } from '../models/video-game';
import { Option } from '../models/option';
import { ApiService } from './api.service';
import { VideoGameUpdateRequest } from '../models/video-game-update-request';
import { CacheService } from './cache.service';

@Injectable({
  providedIn: 'root'
})
export class VideoGameService {
  private readonly VIDEO_GAME_URL = '/api/VideoGame';
  private clearVideoGames = false;

  constructor(private apiService: ApiService,
    private cacheService: CacheService) { }

  public getVideoGames(): Observable<VideoGame[]>{
    const key = "VIDEO_GAMES";
    var res = this.cacheService.tryGet(key, this.apiService.get<VideoGame[]>(this.VIDEO_GAME_URL + '/getVideoGames'), this.clearVideoGames);
    this.clearVideoGames = false;
    return res;
  }

  public getPlatforms(): Observable<Platform[]>{
    const key = "PLATFORMS";
    return  this.cacheService.tryGet(key, this.apiService.get<Platform[]>(this.VIDEO_GAME_URL + '/getPlatforms'));
  }

  public platformListToOptions(platforms: Platform[]): Option[]{
    if(platforms){
      return platforms.map(platform => {
        return {
          key: platform.id,
          value: platform.name
        }
      })
    }
    else{
      return [];
    }
  }

  public getPublishers(): Observable<Publisher[]>{
    const key = "PUBLISHERS";
    return  this.cacheService.tryGet(key, this.apiService.get<Platform[]>(this.VIDEO_GAME_URL + '/getPublishers'));
  }

  public publisherListToOptions(publishers: Publisher[]): Option[]{
    if(publishers){
      return publishers.map(publisher => {
        return {
          key: publisher.id,
          value: publisher.name
        }
      })
    }
    else{
      return [];
    }
  }

  public getGameTypes(): GameType[]{
    let res = [];
    for (let element in GameType) {
      if (!isNaN(Number(element))) {
        res.push(element);
      }
    }
    return res; 
  }

  public gameTypeListToOptions(gameTypes: GameType[]): Option[]{
    if(gameTypes){
      return gameTypes.map(gameType => {
        return {
          key: gameType,
          value: GameType[gameType]
        }
      })
    }
    else{
      return [];
    }
  }

  public save(id: number, name: string, description: string,
    platformOptions: Option[], publisherId: number, gameType: GameType){
    let videoGameUpdateRequest: VideoGameUpdateRequest ={
      id: id,
      name: name,
      description: description,
      platformIds: platformOptions.map(platformOption => platformOption.key as number),
      publisherId: publisherId,
      gameType: gameType
    }
    this.clearVideoGames = true;
    return this.apiService.put<void>(this.VIDEO_GAME_URL + '/update', videoGameUpdateRequest);
  }
}

File: VideoGameServer/Core/Services/VideoGameService.cs
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using VideoGameServer.Core.Models;

namespace VideoGameServer.Core.Services
{
    public class VideoGameService : IVideoGameService
    {
        public IEnumerable<Platform> GetPlatforms(VideoGameDbContext videoGameDbContext)
        {
            return videoGameDbContext.Platforms;
        }

        public IEnumerable<Publisher> GetPublishers(VideoGameDbContext videoGameDbContext)
        {
            return videoGameDbContext.Publishers;
        }

        public IEnumerable<VideoGameResponse> GetVideoGames(VideoGameDbContext videoGameDbContext)
        {
            return videoGameDbContext.VideoGames
                        .Include(x => x.Publisher)
                        .Include(x => x.VideoGamePlatforms)
                        .ThenInclude(x => x.Platform)
                        .Select(x => new VideoGameResponse()
                        {
                            Id = x.Id,
                            Name = x.Name,
                            Description = x.Description,
                            Platforms = x.VideoGamePlatforms.Select(y => y.Platform),
                            Publisher = x.Publisher,
                            GameType = x.GameType
                        });
        }

        public void Update(VideoGameDbContext videoGameDbContext, VideoGameUpdateRequest videoGameUpdateRequest)
        {
            VideoGame videoGame =
                videoGameDbContext
                .VideoGames
                .Include(x => x.VideoGamePlatforms)
                .FirstOrDefault(item => item.Id == videoGameUpdateRequest.Id);
            if (videoGame == null)
            {
                throw new ArgumentException("Cannot find the video game!");
            }
            videoGame.Name = videoGameUpdateRequest.Name;
            videoGame.Description = videoGameUpdateRequest.Description;
            videoGame.GameType = videoGameUpdateRequest.GameType;
            videoGame.PublisherId = videoGameUpdateRequest.PublisherId;

            int[] originalPlatformIds = videoGame.VideoGamePlatforms.Select(x => x.PlatformId).ToArray();
            int[] newPlatformIds = videoGameUpdateRequest.PlatformIds;

            int[] removedPlatformIds = GetRemoved(originalPlatformIds, newPlatformIds);
            int[] addedPlatformIds = GetAdded(originalPlatformIds, newPlatformIds);

            if (removedPlatformIds.Any())
            {
                IEnumerable<VideoGamePlatform> removedVideoGamePlatforms =
                videoGameDbContext
                .VideoGamePlatforms
                .Where(x => removedPlatformIds.Contains(x.PlatformId) && x.VideoGameId == videoGame.Id);

                foreach (VideoGamePlatform videoGamePlatform in removedVideoGamePlatforms)
                {
                    videoGame.VideoGamePlatforms.Remove(videoGamePlatform);
                }
            }

            if (addedPlatformIds.Any())
            {
                foreach (int addedPlatformId in addedPlatformIds)
                {
                    videoGame.VideoGamePlatforms.Add(new VideoGamePlatform()
                    {
                        VideoGameId = videoGame.Id,
                        PlatformId = addedPlatformId
                    });
                }
            }
            try
            {
                videoGameDbContext.VideoGames.Update(videoGame);
                videoGameDbContext.SaveChanges();
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        private int[] GetAdded(int[] originalArray, int[] newArray)
        {
            return newArray.Where(x => !originalArray.Contains(x)).ToArray();
        }

        private int[] GetRemoved(int[] originalArray, int[] newArray)
        {
            return originalArray.Where(x => !newArray.Contains(x)).ToArray();
        }
    }
}
